src/prefect/server/api/concurrency_limits.py

```diff
@@ -188,79 +385,163 @@ async def increment_concurrency_limits_v1(
+        v1_limits = (
+            await concurrency_limits.filter_concurrency_limits_for_orchestration(
+                session, tags=names
+            )
+        )
+        v1_by_tag = {limit.tag: limit for limit in v1_limits}
+
+        # Process each requested tag
+        applied_v2_ids = []
+        applied_v1_tags = []
+        results = []
+
+        for tag in names:
+            v2_limit = v2_by_name.get(f"tag:{tag}")
+            v1_limit = v1_by_tag.get(tag)
+
+            # Prefer V2 if it exists
+            TAG_CONCURRENCY_SLOT_WAIT_SECONDS = (
+                get_current_settings().server.tasks.tag_concurrency_slot_wait_seconds
             )
-            run_limits = {limit.tag: limit for limit in filtered_limits}
-            for tag, cl in run_limits.items():
-                limit = cl.concurrency_limit
-                if limit == 0:
-                    # limits of 0 will deadlock, and the transition needs to abort
-                    for stale_tag in applied_limits.keys():
-                        stale_limit = run_limits.get(stale_tag, None)
-                        active_slots = set(stale_limit.active_slots)
+            if v2_limit:
+                # Check for zero limit
+                if v2_limit.limit == 0:
+                    # Roll back any already applied limits
+                    if applied_v2_ids:
+                        await cl_v2_models.bulk_decrement_active_slots(
+                            session=session,
+                            concurrency_limit_ids=applied_v2_ids,
+                            slots=1,
+                        )
+                    for applied_tag in applied_v1_tags:
+                        cl = v1_by_tag[applied_tag]
+                        active_slots = set(cl.active_slots)
                         active_slots.discard(str(task_run_id))
-                        stale_limit.active_slots = list(active_slots)
+                        cl.active_slots = list(active_slots)
 
-                    raise Abort(
-                        reason=(
+                    raise HTTPException(
+                        status_code=status.HTTP_423_LOCKED,
+                        detail=(
                             f'The concurrency limit on tag "{tag}" is 0 and will '
                             "deadlock if the task tries to run again."
                         ),
                     )
-                elif len(cl.active_slots) >= limit:
-                    # if the limit has already been reached, delay the transition
-                    for stale_tag in applied_limits.keys():
-                        stale_limit = run_limits.get(stale_tag, None)
-                        active_slots = set(stale_limit.active_slots)
+
+                # Try to acquire V2 slot
+                if v2_limit.active:
+                    acquired = await cl_v2_models.bulk_increment_active_slots(
+                        session=session,
+                        concurrency_limit_ids=[v2_limit.id],
```

Comment:
Why are we only acquiring one limit at a time here? I think we should try to get all of them at once so that we don't have to worry about rollback.